File: src/screens/OnboardingScreen.js
import React, { useState } from 'react';
import { View, Text, StyleSheet, TouchableOpacity, Alert, Linking, Platform, ScrollView } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { LinearGradient } from 'expo-linear-gradient';
import * as Location from 'expo-location';
import * as Notifications from 'expo-notifications';
import { Colors, BorderRadius, Spacing } from '../constants/theme';

const STEPS = [
  {
    key: 'welcome',
    icon: 'notifications',
    iconBg: Colors.secondaryContainer,
    iconColor: Colors.onSecondaryContainer,
    title: 'Vigilant Path',
    subtitle: 'Never miss your stop again.',
    description: 'Set location-based alarms that automatically alert you with sound and vibration when you arrive at your destination.',
    features: [
      { icon: 'location', text: 'GPS Geofencing' },
      { icon: 'volume-high', text: 'Alarm Sound' },
      { icon: 'phone-portrait', text: 'Vibration' },
      { icon: 'moon', text: 'Snooze & Re-trigger' },
    ],
    buttonText: 'Get Started',
  },
  {
    key: 'location',
    icon: 'location',
    iconBg: Colors.primaryFixed,
    iconColor: Colors.primary,
    title: 'Location Access',
    subtitle: 'Required for alarms to work',
    description: 'We need your location to detect when you\'re approaching your destination and trigger the alarm automatically.',
    details: [
      { icon: 'shield-checkmark', text: 'Your location data stays on your device only' },
      { icon: 'cloud-offline', text: 'No location data is sent to any server' },
      { icon: 'battery-charging', text: 'Smart battery optimization - uses GPS only when needed' },
    ],
    buttonText: 'Allow Location Access',
  },
  {
    key: 'background',
    icon: 'navigate-circle',
    iconBg: '#fef3c7',
    iconColor: '#d97706',
    title: '"Allow All The Time"',
    subtitle: 'Essential for background monitoring',
    description: 'For alarms to work when the app is in the background or your phone is locked, you MUST select "Allow all the time" in the next prompt.',
    details: [
      { icon: 'alert-circle', text: 'Choose "Allow all the time" — NOT "While using the app"', bold: true },
      { icon: 'alarm', text: 'Without this, alarms won\'t trigger when the screen is off' },
      { icon: 'information-circle', text: 'You can change this later in Settings → Apps → Vigilant Path' },
    ],
    buttonText: 'Grant Background Access',
    warningText: '⚠️ Select "Allow all the time" on the next screen',
  },
  {
    key: 'notifications',
    icon: 'notifications-circle',
    iconBg: Colors.tertiaryFixed,
    iconColor: Colors.tertiary,
    title: 'Notifications',
    subtitle: 'Get alerted when you arrive',
    description: 'We\'ll send you a notification when you enter your alarm\'s geofence, so you\'ll be alerted even if the app is in the background.',
    details: [
      { icon: 'megaphone', text: 'Alarm trigger notifications' },
      { icon: 'pulse', text: 'Background monitoring status' },
    ],
    buttonText: 'Allow Notifications',
  },
];

export default function OnboardingScreen({ navigation, onComplete }) {
  const insets = useSafeAreaInsets();
  const [currentStep, setCurrentStep] = useState(0);

  const handleNext = async () => {
    const step = STEPS[currentStep];

    if (step.key === 'welcome') {
      setCurrentStep(1);
      return;
    }

    if (step.key === 'location') {
      const { status } = await Location.requestForegroundPermissionsAsync();
      if (status !== 'granted') {
        Alert.alert(
          'Location Required',
          'Location access is essential for this app to work. Please grant location permission.',
          [
            { text: 'Open Settings', onPress: () => Linking.openSettings() },
            { text: 'Try Again', onPress: () => handleNext() },
          ]
        );
        return;
      }
      setCurrentStep(2);
      return;
    }

    if (step.key === 'background') {
      const { status } = await Location.requestBackgroundPermissionsAsync();
      if (status !== 'granted') {
        Alert.alert(
          '⚠️ "Allow All The Time" Required',
          'You need to select "Allow all the time" for alarms to work in the background.\n\nPlease go to Settings and change the location permission to "Allow all the time".',
          [
            { text: 'Open Settings', onPress: () => Linking.openSettings() },
            {
              text: 'Try Again',
              onPress: async () => {
                // Check again after returning from settings
                const recheck = await Location.getBackgroundPermissionsAsync();
                if (recheck.status === 'granted') {
                  setCurrentStep(3);
                } else {
                  Alert.alert(
                    'Still Not Granted',
                    'Background location is still not set to "Allow all the time". Alarms may not work properly in the background.\n\nYou can change this later in phone Settings.',
                    [
                      { text: 'Continue Anyway', onPress: () => setCurrentStep(3) },
                      { text: 'Open Settings', onPress: () => Linking.openSettings() },
                    ]
                  );
                }
              },
            },
          ]
        );
        return;
      }
      setCurrentStep(3);
      return;
    }

    if (step.key === 'notifications') {
      await Notifications.requestPermissionsAsync();
      // Complete onboarding regardless of notification permission
      onComplete();
      return;
    }
  };

  const step = STEPS[currentStep];

  return (
    <View style={[styles.container, { paddingTop: insets.top + 24, paddingBottom: insets.bottom + 24 }]}>
      {/* Progress dots */}
      <View style={styles.progressRow}>
        {STEPS.map((_, i) => (
          <View
            key={i}
            style={[
              styles.progressDot,
              i === currentStep && styles.progressDotActive,
              i < currentStep && styles.progressDotDone,
            ]}
          />
        ))}
      </View>

      <ScrollView showsVerticalScrollIndicator={false} contentContainerStyle={styles.scrollContent}>
        {/* Icon */}
        <View style={styles.heroSection}>
          <View style={[styles.iconCircle, { backgroundColor: step.iconBg }]}>
            <Ionicons name={step.icon} size={48} color={step.iconColor} />
          </View>
        </View>

        {/* Title */}
        <Text style={styles.title}>{step.title}</Text>
        <Text style={styles.subtitle}>{step.subtitle}</Text>
        <Text style={styles.description}>{step.description}</Text>

        {/* Features (welcome screen) */}
        {step.features && (
          <View style={styles.featuresGrid}>
            {step.features.map((f, i) => (
              <View key={i} style={styles.featureItem}>
                <View style={styles.featureIcon}>
                  <Ionicons name={f.icon} size={20} color={Colors.primary} />
                </View>
                <Text style={styles.featureText}>{f.text}</Text>
              </View>
            ))}
          </View>
        )}

        {/* Details list (permission screens) */}
        {step.details && (
          <View style={styles.detailsList}>
            {step.details.map((d, i) => (
              <View key={i} style={styles.detailItem}>
                <Ionicons name={d.icon} size={20} color={d.bold ? '#d97706' : Colors.primary} />
                <Text style={[styles.detailText, d.bold && styles.detailTextBold]}>{d.text}</Text>
              </View>
            ))}
          </View>
        )}

        {/* Warning text */}
        {step.warningText && (
          <View style={styles.warningBox}>
            <Text style={styles.warningText}>{step.warningText}</Text>
          </View>
        )}
      </ScrollView>

      {/* CTA Button */}
      <TouchableOpacity
        activeOpacity={0.85}
        onPress={handleNext}
        style={styles.ctaWrapper}
      >
        <LinearGradient
          colors={[Colors.primary, Colors.primaryContainer]}
          start={{ x: 0, y: 0 }}
          end={{ x: 1, y: 1 }}
          style={styles.ctaButton}
        >
          <Text style={styles.ctaText}>{step.buttonText}</Text>
          <Ionicons name="arrow-forward" size={20} color="#fff" />
        </LinearGradient>
      </TouchableOpacity>

      {/* Skip option (only on notification step) */}
      {step.key === 'notifications' && (
        <TouchableOpacity onPress={onComplete} style={styles.skipBtn}>
          <Text style={styles.skipText}>Skip for now</Text>
        </TouchableOpacity>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.surfaceContainerLow,
    paddingHorizontal: Spacing.xl,
  },
  progressRow: {
    flexDirection: 'row',
    justifyContent: 'center',
    gap: 8,
    marginBottom: Spacing.xl,
  },
  progressDot: {
    width: 8, height: 8, borderRadius: 4,
    backgroundColor: Colors.surfaceContainerHighest,
  },
  progressDotActive: {
    width: 28, backgroundColor: Colors.primary,
  },
  progressDotDone: {
    backgroundColor: Colors.primaryFixedDim,
  },
  scrollContent: {
    flexGrow: 1,
  },
  heroSection: {
    alignItems: 'center',
    paddingVertical: Spacing.xxl,
  },
  iconCircle: {
    width: 100, height: 100, borderRadius: 50,
    alignItems: 'center', justifyContent: 'center',
    shadowColor: '#000', shadowOffset: { width: 0, height: 4 },
    shadowOpacity: 0.12, shadowRadius: 16, elevation: 6,
  },
  title: {
    fontSize: 32, fontWeight: '800', color: Colors.onSurface,
    letterSpacing: -0.5, textAlign: 'center',
  },
  subtitle: {
    fontSize: 15, fontWeight: '600', color: Colors.primary,
    textAlign: 'center', marginTop: 4,
  },
  description: {
    fontSize: 15, color: Colors.onSurfaceVariant,
    lineHeight: 22, textAlign: 'center',
    marginTop: Spacing.lg, marginBottom: Spacing.xl,
    paddingHorizontal: Spacing.md,
  },
  featuresGrid: {
    flexDirection: 'row', flexWrap: 'wrap',
    justifyContent: 'center', gap: Spacing.md,
    marginBottom: Spacing.xl,
  },
  featureItem: {
    flexDirection: 'row', alignItems: 'center', gap: 8,
    backgroundColor: Colors.primaryFixed,
    paddingHorizontal: 14, paddingVertical: 10,
    borderRadius: BorderRadius.full,
  },
  featureIcon: {},
  featureText: { fontSize: 12, fontWeight: '700', color: Colors.primary },
  detailsList: {
    gap: Spacing.md, marginBottom: Spacing.xl,
  },
  detailItem: {
    flexDirection: 'row', alignItems: 'flex-start', gap: Spacing.md,
    backgroundColor: Colors.surfaceContainerHigh,
    padding: Spacing.lg, borderRadius: BorderRadius.xl,
  },
  detailText: {
    flex: 1, fontSize: 14, color: Colors.onSurfaceVariant, lineHeight: 20,
  },
  detailTextBold: {
    fontWeight: '700', color: '#92400e',
  },
  warningBox: {
    backgroundColor: '#fef3c7', padding: Spacing.lg,
    borderRadius: BorderRadius.xl, borderWidth: 1,
    borderColor: '#fbbf24', marginBottom: Spacing.lg,
  },
  warningText: {
    fontSize: 14, fontWeight: '700', color: '#92400e', textAlign: 'center',
  },
  ctaWrapper: {
    borderRadius: BorderRadius.full, overflow: 'hidden',
    shadowColor: Colors.primary, shadowOffset: { width: 0, height: 8 },
    shadowOpacity: 0.25, shadowRadius: 24, elevation: 8,
    marginTop: Spacing.lg,
  },
  ctaButton: {
    flexDirection: 'row', alignItems: 'center', justifyContent: 'center',
    gap: Spacing.sm, paddingVertical: 18, borderRadius: BorderRadius.full,
  },
  ctaText: { color: '#fff', fontSize: 18, fontWeight: '700' },
  skipBtn: {
    alignItems: 'center', paddingVertical: Spacing.md, marginTop: Spacing.sm,
  },
  skipText: { fontSize: 14, color: Colors.outline, fontWeight: '600' },
});

File: android/app/src/main/java/com/vigilantpath/locationbasedalarm/ui/screens/OnboardingScreen.kt
package com.vigilantpath.locationbasedalarm.ui.screens

import android.Manifest
import android.os.Build
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.vigilantpath.locationbasedalarm.data.PreferencesManager
import com.vigilantpath.locationbasedalarm.ui.theme.*
import kotlinx.coroutines.launch

data class OnboardingStep(
    val key: String,
    val icon: ImageVector,
    val iconBg: Color,
    val iconColor: Color,
    val title: String,
    val subtitle: String,
    val description: String,
    val features: List<Pair<ImageVector, String>>? = null,
    val details: List<Triple<ImageVector, String, Boolean>>? = null,
    val warningText: String? = null,
    val buttonText: String,
)

private val steps = listOf(
    OnboardingStep(
        key = "welcome",
        icon = Icons.Default.Notifications,
        iconBg = SecondaryFixed,
        iconColor = Secondary,
        title = "Vigilant Path",
        subtitle = "Never miss your stop again.",
        description = "Set location-based alarms that automatically alert you with sound and vibration when you arrive at your destination.",
        features = listOf(
            Icons.Default.LocationOn to "GPS Geofencing",
            Icons.Default.VolumeUp to "Alarm Sound",
            Icons.Default.PhoneAndroid to "Vibration",
            Icons.Default.DarkMode to "Snooze & Re-trigger",
        ),
        buttonText = "Get Started",
    ),
    OnboardingStep(
        key = "location",
        icon = Icons.Default.LocationOn,
        iconBg = PrimaryFixed,
        iconColor = Primary,
        title = "Location Access",
        subtitle = "Required for alarms to work",
        description = "We need your location to detect when you're approaching your destination and trigger the alarm automatically.",
        details = listOf(
            Triple(Icons.Default.Shield, "Your location data stays on your device only", false),
            Triple(Icons.Default.CloudOff, "No location data is sent to any server", false),
            Triple(Icons.Default.BatteryChargingFull, "Smart battery optimization", false),
        ),
        buttonText = "Allow Location Access",
    ),
    OnboardingStep(
        key = "background",
        icon = Icons.Default.Explore,
        iconBg = WarningBg,
        iconColor = WarningAccent,
        title = "\"Allow All The Time\"",
        subtitle = "Essential for background monitoring",
        description = "For alarms to work in the background, you MUST select \"Allow all the time\" in the next prompt.",
        details = listOf(
            Triple(Icons.Default.Warning, "Choose \"Allow all the time\" — NOT \"While using the app\"", true),
            Triple(Icons.Default.Alarm, "Without this, alarms won't trigger when screen is off", false),
            Triple(Icons.Default.Info, "You can change this later in Settings → Apps", false),
        ),
        warningText = "⚠️ Select \"Allow all the time\" on the next screen",
        buttonText = "Grant Background Access",
    ),
    OnboardingStep(
        key = "notifications",
        icon = Icons.Default.NotificationsActive,
        iconBg = TertiaryFixed,
        iconColor = Tertiary,
        title = "Notifications",
        subtitle = "Get alerted when you arrive",
        description = "We'll send you a notification when you enter your alarm's geofence, even if the app is in the background.",
        details = listOf(
            Triple(Icons.Default.Campaign, "Alarm trigger notifications", false),
            Triple(Icons.Default.FavoriteBorder, "Background monitoring status", false),
        ),
        buttonText = "Allow Notifications",
    ),
)

@OptIn(androidx.compose.foundation.layout.ExperimentalLayoutApi::class)
@Composable
fun OnboardingScreen(onComplete: () -> Unit) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val prefs = remember { PreferencesManager(context) }
    var currentStep by remember { mutableIntStateOf(0) }
    val step = steps[currentStep]

    val locationPermissionLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { permissions ->
        if (permissions[Manifest.permission.ACCESS_FINE_LOCATION] == true) {
            currentStep = 2
        }
    }

    val backgroundPermissionLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { _ -> currentStep = 3 }

    val notificationPermissionLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { _ ->
        scope.launch {
            prefs.setOnboardingDone(true)
            onComplete()
        }
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(SurfaceContainerLow)
            .statusBarsPadding()
            .navigationBarsPadding()
            .padding(horizontal = 24.dp)
            .padding(top = 12.dp, bottom = 16.dp)
    ) {
        // Progress dots
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.Center
        ) {
            steps.forEachIndexed { i, _ ->
                Box(
                    modifier = Modifier
                        .padding(horizontal = 4.dp)
                        .size(
                            width = if (i == currentStep) 28.dp else 8.dp,
                            height = 8.dp
                        )
                        .clip(CircleShape)
                        .background(
                            when {
                                i == currentStep -> Primary
                                i < currentStep -> PrimaryFixedDim
                                else -> SurfaceContainerHighest
                            }
                        )
                )
            }
        }

        // Scrollable content area
        Column(
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState()),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Spacer(Modifier.height(16.dp))

            // Icon - smaller to save space
            Box(
                modifier = Modifier
                    .size(72.dp)
                    .shadow(4.dp, CircleShape)
                    .clip(CircleShape)
                    .background(step.iconBg),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    imageVector = step.icon,
                    contentDescription = null,
                    tint = step.iconColor,
                    modifier = Modifier.size(36.dp)
                )
            }

            Spacer(Modifier.height(16.dp))

            // Title
            Text(
                text = step.title,
                fontSize = 28.sp,
                fontWeight = FontWeight.ExtraBold,
                color = OnSurface,
                textAlign = TextAlign.Center,
            )

            Text(
                text = step.subtitle,
                style = MaterialTheme.typography.labelLarge,
                color = Primary,
                textAlign = TextAlign.Center,
                modifier = Modifier.padding(top = 2.dp)
            )

            Text(
                text = step.description,
                style = MaterialTheme.typography.bodySmall,
                color = OnSurfaceVariant,
                textAlign = TextAlign.Center,
                lineHeight = 18.sp,
                modifier = Modifier.padding(top = 10.dp, start = 8.dp, end = 8.dp)
            )

            Spacer(Modifier.height(16.dp))

            // Features (welcome screen)
            step.features?.let { features ->
                FlowRow(
                    horizontalArrangement = Arrangement.Center,
                    modifier = Modifier.fillMaxWidth()
                ) {
                    features.forEach { (icon, text) ->
                        Row(
                            modifier = Modifier
                                .padding(4.dp)
                                .background(PrimaryFixed, CircleShape)
                                .padding(horizontal = 12.dp, vertical = 8.dp),
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(6.dp)
                        ) {
                            Icon(icon, null, tint = Primary, modifier = Modifier.size(16.dp))
                            Text(text, fontSize = 11.sp, fontWeight = FontWeight.Bold, color = Primary)
                        }
                    }
                }
            }

            // Details (permission screens) - compact
            step.details?.let { details ->
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    details.forEach { (icon, text, bold) ->
                        Row(
                            modifier = Modifier
                                .fillMaxWidth()
                                .background(SurfaceContainerHigh, RoundedCornerShape(16.dp))
                                .padding(12.dp),
                            horizontalArrangement = Arrangement.spacedBy(10.dp),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Icon(
                                icon, null,
                                tint = if (bold) WarningAccent else Primary,
                                modifier = Modifier.size(18.dp)
                            )
                            Text(
                                text = text,
                                fontSize = 13.sp,
                                lineHeight = 17.sp,
                                color = if (bold) WarningText else OnSurfaceVariant,
                                fontWeight = if (bold) FontWeight.Bold else FontWeight.Normal,
                                modifier = Modifier.weight(1f)
                            )
                        }
                    }
                }
            }

            // Warning - compact
            step.warningText?.let { warning ->
                Spacer(Modifier.height(10.dp))
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .background(WarningBg, RoundedCornerShape(16.dp))
                        .padding(12.dp)
                ) {
                    Text(
                        text = warning,
                        fontSize = 13.sp,
                        color = WarningText,
                        textAlign = TextAlign.Center,
                        fontWeight = FontWeight.Bold,
                        modifier = Modifier.fillMaxWidth()
                    )
                }
            }

            Spacer(Modifier.height(16.dp))
        }

        // CTA Button - always visible at bottom
        Button(
            onClick = {
                when (step.key) {
                    "welcome" -> currentStep = 1
                    "location" -> {
                        locationPermissionLauncher.launch(
                            arrayOf(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION)
                        )
                    }
                    "background" -> {
                        backgroundPermissionLauncher.launch(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
                    }
                    "notifications" -> {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                            notificationPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
                        } else {
                            scope.launch { prefs.setOnboardingDone(true); onComplete() }
                        }
                    }
                }
            },
            modifier = Modifier
                .fillMaxWidth()
                .height(52.dp),
            shape = CircleShape,
            colors = ButtonDefaults.buttonColors(containerColor = Color.Transparent),
            contentPadding = PaddingValues(0.dp)
        ) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .background(Brush.linearGradient(listOf(Primary, PrimaryContainer)), CircleShape),
                contentAlignment = Alignment.Center
            ) {
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    Text(step.buttonText, fontSize = 16.sp, fontWeight = FontWeight.Bold, color = Color.White)
                    Icon(Icons.AutoMirrored.Filled.ArrowForward, null, tint = Color.White, modifier = Modifier.size(18.dp))
                }
            }
        }

        if (step.key == "notifications") {
            TextButton(
                onClick = { scope.launch { prefs.setOnboardingDone(true); onComplete() } },
                modifier = Modifier.fillMaxWidth()
            ) {
                Text("Skip for now", color = Outline, fontWeight = FontWeight.SemiBold, fontSize = 13.sp)
            }
        }
    }
}
